### PR TITLE
Duplicated profiles

### PR DESCRIPTION
**Pull request by godrei**

### CODE REVIEW

File: lib/autoprovision/portal/profile_client.rb [hunk 1]
```diff
@@ -25,57 +25,114 @@ def self.ensure_xcode_managed_profile(bundle_id, entitlements, distribution_type
...
+      end
+
       begin
-        Log.debug("generating #{distribution_type} profile: #{profile_name}")
+        Log.debug("generating profile: #{profile_name}")
         profile_class = portal_profile_class(distribution_type)
         run_and_handle_portal_function { profile = profile_class.create!(bundle_id: app.bundle_id, certificate: certificate, name: profile_name, sub_platform: platform == :tvos ? 'tvOS' : nil) }
       rescue => ex
         raise ex unless allow_retry
         raise ex unless ex.to_s =~ /Multiple profiles found with the name/i
```
Comment (BirmacherAkos) [inline lib/autoprovision/portal/profile_client.rb]: Typo?: `/i`

Comment (godrei) [reply lib/autoprovision/portal/profile_client.rb]: insensitive flag for regex: https://stackoverflow.com/a/41266812

File: lib/autoprovision/portal/profile_client.rb [hunk 2]
```diff
@@ -25,57 +25,114 @@ def self.ensure_xcode_managed_profile(bundle_id, entitlements, distribution_type
...
     def self.ensure_manual_profile(certificate, app, entitlements, distribution_type, platform, min_profile_days_valid, allow_retry = true)
+      all_profiles = ProfileClient.fetch_profiles(false, platform)
+
+      # search for the Bitrise managed profile
       profile_name = "Bitrise #{distribution_type} - (#{app.bundle_id})"
+      profile = all_profiles.select { |prof| prof.name == profile_name }.first
+      valid = !profile.nil?
+
+      # check the profile's bundle id
+      if valid
```
Comment (trapacska) [inline lib/autoprovision/portal/profile_client.rb]: These ifs could be merged.

File: lib/autoprovision/portal/profile_client.rb [hunk 3]
```diff
@@ -25,57 +25,114 @@ def self.ensure_xcode_managed_profile(bundle_id, entitlements, distribution_type
     end
 
     def self.ensure_manual_profile(certificate, app, entitlements, distribution_type, platform, min_profile_days_valid, allow_retry = true)
+      all_profiles = ProfileClient.fetch_profiles(false, platform)
+
+      # search for the Bitrise managed profile
       profile_name = "Bitrise #{distribution_type} - (#{app.bundle_id})"
+      profile = all_profiles.select { |prof| prof.name == profile_name }.first
```
Comment (trapacska) [inline lib/autoprovision/portal/profile_client.rb]: why the first only?

File: lib/autoprovision/portal/profile_client.rb [hunk 4]
```diff
@@ -25,57 +25,114 @@ def self.ensure_xcode_managed_profile(bundle_id, entitlements, distribution_type
...
 
-      profile = nil
+      return profile if valid
+
+      # profile name needs to be unique
+      if !valid && !profile.nil?
+        profile.delete!
+        ProfileClient.clear_cache(false, platform)
+      end
+
```
Comment (trapacska) [inline lib/autoprovision/portal/profile_client.rb]: from here to the top is a very long logic, I would put them in functions. For example: isBundleIDValid, isDistributionTypeValid, isCapabilitiesValid

File: lib/autoprovision/portal/profile_client.rb [hunk 5]
```diff
@@ -25,57 +25,114 @@ def self.ensure_xcode_managed_profile(bundle_id, entitlements, distribution_type
...
+      # search for the Bitrise managed profile
       profile_name = "Bitrise #{distribution_type} - (#{app.bundle_id})"
+      profile = all_profiles.select { |prof| prof.name == profile_name }.first
+      valid = !profile.nil?
+
+      # check the profile's bundle id
+      if valid
+        if profile.app.bundle_id != app.bundle_id
+          Log.debug("Profile (#{profile.name}) bundle id: #{profile.app.bundle_id}, should be: #{app.bundle_id}")
```
Comment (lpusok) [inline lib/autoprovision/portal/profile_client.rb]: Is debug log on by default?

**Comment by lpusok** [comment]
👍

**Comment by lszucs** [comment]
Nothing to add to previous comments.

File: lib/autoprovision/portal/profile_client.rb (hunk 3 above)
Comment (godrei) [reply lib/autoprovision/portal/profile_client.rb]: as the name is unique on DevPortal

File: lib/autoprovision/portal/profile_client.rb (hunk 5 above)
Comment (godrei) [reply lib/autoprovision/portal/profile_client.rb]: nope

File: lib/autoprovision/portal/profile_client.rb [hunk 6]
```diff
@@ -7,155 +7,152 @@ module Portal
...
+        'ad-hoc' => 'adhoc',
+        'enterprise' => 'inhouse'
+      }
+      desired_distribution_method = distribution_methods[distribution_type]
+
+      unless profile.distribution_method == desired_distribution_method
+        Log.debug("Profile (#{profile.name}) distribution type: #{profile.distribution_method}, should be: #{desired_distribution_method}")
+        return false
       end
+      false
```
Comment (BirmacherAkos) [inline lib/autoprovision/portal/profile_client.rb]: This should be true.

File: lib/autoprovision/portal/profile_client.rb [hunk 7]
```diff
@@ -7,155 +7,152 @@ module Portal
...
+
+        matching_profiles.push(profile) if File.fnmatch(profile.app.bundle_id, bundle_id)
       end
 
-      profile
-    end
+      profiles = full_matching_profiles.select do |profile|
+        distribution_type_matches?(profile, distribution_type) &&
+          !expired?(profile, min_profile_days_valid) &&
+          all_services_enabled?(entitlements, entitlements) &&
```
Comment (BirmacherAkos) [inline lib/autoprovision/portal/profile_client.rb]: typo:
it should be (profile, entitlements)

You use it wrong in the other cases as well in this file! Check them as well.

File: lib/autoprovision/portal/profile_client.rb (hunk 6 above)
Comment (BirmacherAkos) [reply lib/autoprovision/portal/profile_client.rb]: Resolved in:
https://github.com/bitrise-steplib/steps-ios-auto-provision/pull/59/commits/f2df743bed708ae868c6ea36486a30820ef81888

File: lib/autoprovision/portal/profile_client.rb (hunk 7 above)
Comment (BirmacherAkos) [reply lib/autoprovision/portal/profile_client.rb]: Resolved in:
https://github.com/bitrise-steplib/steps-ios-auto-provision/pull/59/commits/f2df743bed708ae868c6ea36486a30820ef81888

**Pull request merged**